### PR TITLE
AWS EMR Module

### PR DESCRIPTION
**Pull request by ferrants**
Module for managing EMR clusters. We have been using this for months and would like to contribute it back.

@mbomio FYI

### CODE REVIEW

**Comment by mbomio** [comment]
:+1:

File: cloud/amazon/emr.py [hunk 1]
```diff
@@ -0,0 +1,420 @@
...
+# (at your option) any later version.
+#
+# Ansible is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Ansible.  If not, see <http://www.gnu.org/licenses/>.
+
```
Comment (bcoca) [inline cloud/amazon/emr.py:116]: missing options documentation and return docs

File: cloud/amazon/emr.py [hunk 2]
```diff
@@ -0,0 +1,420 @@
...
+from datetime import datetime
+
+try:
+    import boto.emr
+    from boto.emr.connection import EmrConnection
+    from boto.emr.instance_group import InstanceGroup
+    from boto.emr.step import JarStep
+    from boto.emr.step import InstallHiveStep
+    from boto.emr.bootstrap_action import BootstrapAction
+except ImportError:
```
Comment (bcoca) [inline cloud/amazon/emr.py:261]: modules should not exit, set a variable and use the json_fail, look at other aws modules for an example (HAS_BOTO)

**Comment by mbomio** [comment]
@bcoca I added the options.

**Comment by bcoca** [comment]
cc @defionscode

**Comment by dan-coates** [comment]
What's the status of this pull request? Can this be merged? An EMR module would be useful for me and many others I'm sure.

**Comment by ferrants** [comment]
It works for me and @mbomio and can probably be merged. If you want to use it before it is merged, you could copy the file into your ansible directory's library folder or wherever you have your custom modules at.

**Comment by pmackay** [comment]
I'd also really appreciate this being merged.

**Comment by joshmullis** [comment]
ditto

**Comment by ferrants** [comment]
I'll leave this open, but ansible is really not that great at merging in new features, I wouldn't count on it being merged any time soon. If you want to begin using this, I think you should copy this file into your `library/` directory so you can use it.

File: cloud/amazon/emr.py [hunk 3]
```diff
@@ -0,0 +1,522 @@
...
+def main():
+    module = AnsibleModule(
+        argument_spec=dict(
+            ec2_keyname={'default': None},
+            cluster_name={},
+            jobflow_id={'default': None},
+            log_uri={},
+            ec2_secret_key={'default': None, 'aliases': ['aws_secret_key', 'secret_key'], 'no_log': True},
+            ec2_access_key={'default': None, 'aliases': ['aws_access_key', 'access_key']},
+            region={'default': None, 'aliases': ['aws_region', 'ec2_region'], 'choices': AWS_REGIONS},
```
Comment (Sabbasth) [inline cloud/amazon/emr.py:477]: Where is `AWS_REGIONS` defined ? I have an error about this when trying to use the module

```
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: NameError: global name 'AWS_REGIONS' is not defined
fatal: [localhost]: FAILED! => {"changed": false, "failed": true, "module_stderr": "Traceback (most recent call last):
  File "/var/folders/ll/by_nmkg14ng_kvyn_wty1dlc0000gn/T/ansible_qedzmd/ansible_module_emr.py", line 526, in <module>
    main()
  File "/var/folders/ll/by_nmkg14ng_kvyn_wty1dlc0000gn/T/ansible_qedzmd/ansible_module_emr.py", line 483, in main
    region={'default': None, 'aliases': ['aws_region', 'ec2_region'], 'choices': AWS_REGIONS},
NameError: global name 'AWS_REGIONS' is not defined
", "module_stdout": "", "msg": "MODULE FAILURE", "parsed": false}
```

**Comment by ferrants** [comment]
soo, modules are going back into ansible/ansible now? You guys are a mess

**Pull request closed without merging**